File: python/src/threads/telemetry.py
"""The telemetry protocol (spec/api.json `Exporter`): what `host(telemetry=...)` calls.
`threads.otel.otel()` implements it; core only knows the shape and which store a host bound it
to."""

from dataclasses import dataclass
from typing import Literal, Protocol
from weakref import WeakKeyDictionary

from threads.agents.store import Store
from threads.log import BranchId, ThreadId
from threads.result import Err, Ok


@dataclass(frozen=True, slots=True)
class SkippedBranch:
    """A branch `Exporter.sync()` could not read, and the head it failed at."""

    branch_id: BranchId
    thread_id: ThreadId
    code: str
    """Why it didn't read: a log error code such as log_corrupt."""
    head_seq: int


@dataclass(frozen=True, slots=True)
class SyncReport:
    """What one `Exporter.sync()` sent."""

    spans: int
    possibly_lost_events: int
    skipped: tuple[SkippedBranch, ...]


@dataclass(frozen=True, slots=True)
class SyncError:
    code: Literal["collector_unavailable", "collector_rejected"]
    message: str
    status: int | None = None
    """The collector's HTTP status, when it answered."""


class Exporter(Protocol):
    """A telemetry exporter, such as `otel()`."""

    async def sync(self) -> Ok[SyncReport] | Err[SyncError]:
        """Sends every span that closed since the last acknowledged sync."""
        ...


_BOUND: "WeakKeyDictionary[Exporter, Store]" = WeakKeyDictionary()


def bind_telemetry(exporter: Exporter, store: Store) -> None:
    """`host(store=..., telemetry=...)`: an exporter made without a store exports the host's."""
    _BOUND[exporter] = store


def bound_store(exporter: Exporter) -> Store | None:
    """The store a host bound the exporter to, if any."""
    return _BOUND.get(exporter)

File: typescript/packages/core/src/telemetry.ts
import type { Store } from "./agent/sqlite";
import type { BranchId, ThreadId } from "./log";
import type { Result } from "./result";
import type { LogError } from "./verify/error";

// The telemetry protocol (spec/api.json Exporter): what host({telemetry}) calls. otel() in
// @threads/otel implements it; core only knows the shape and which store a host bound it to.

/** A branch Exporter.sync() could not read, and the head it failed at. */
export type SkippedBranch = {
  readonly branch_id: BranchId;
  readonly thread_id: ThreadId;
  readonly code: LogError["code"];
  readonly head_seq: number;
};

/** What one Exporter.sync() sent. */
export type SyncReport = {
  readonly spans: number;
  readonly possiblyLostEvents: number;
  readonly skipped: readonly SkippedBranch[];
};

export type SyncError = {
  readonly code: "collector_unavailable" | "collector_rejected";
  /** The collector's HTTP status, when it answered. */
  readonly status?: number;
  readonly message: string;
};

/** A telemetry exporter, such as otel(). */
export type Exporter = {
  readonly sync: () => Promise<Result<SyncReport, SyncError>>;
};

/** What a host gives its exporter: its store, and the signal its stop() aborts. */
export type TelemetryBinding = {
  readonly store: Store;
  readonly signal: AbortSignal;
};

const BOUND = new WeakMap<Exporter, TelemetryBinding>();

/**
 * host({store, telemetry}): an exporter made without a store exports the host's, and once the
 * host's stop() aborts `signal` it sends nothing more and moves no cursor.
 */
export function bindTelemetry(
  exporter: Exporter,
  binding: TelemetryBinding,
): void {
  BOUND.set(exporter, binding);
}

/** What a host bound the exporter to, if anything. */
export function telemetryBinding(
  exporter: Exporter,
): TelemetryBinding | undefined {
  return BOUND.get(exporter);
}
